Add support for Tuya Wireless Switch entity

lanthaler opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change

This change adds support for for Tuya wireless switches such as the one depicted below with which this change has been tested.

<img src="https://github.com/user-attachments/assets/b703f7ae-90aa-48e9-aec6-685db564a573" width="300">

![tuya-device-page](https://github.com/user-attachments/assets/4097200c-ce32-4e76-bee9-f79e82df827f)

This required a change in tuya-device-sharing-sdk. The relevant [pull request](https://github.com/tuya/tuya-device-sharing-sdk/pull/17) has been merged and released in a new version. Here's the [changelog for tuya-device-sharing-sdk on PyPI](https://pypi.org/project/tuya-device-sharing-sdk/#description:~:text=Release%20Note,updated_status_properties%20to%20SharingDeviceListener) and here's a [diff of all the code changes between the two versions](https://github.com/tuya/tuya-device-sharing-sdk/compare/1496fc3c24643d0045e722698cd34978ce2636d1..e3da82c3558988736b9a8b6df3cb810a4cb7eb1b)


## Type of change

- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [X] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue:
  - fixes #120460 
  - fixes #120458
  - fixes #120020
- This PR is related to issue:
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/34174

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [X] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [x] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

frenck commented on homeassistant/components/tuya/device_trigger.py:
```diff
@@ -0,0 +1,86 @@
+"""Provides device triggers for Tuya."""
```
We should not add device triggers for these, as this is provided for the `event` entities.

Let's remove them.

lanthaler replied:
Thanks for the quick review @frenck. I'm not sure I understand your comment. The built-in Home Assistant Event component doesn't configure any device triggers itself as it doesn't understand the semantics of the events unfortunately in the current form. This configures them just for the Tuya event entities, in this case the wireless switch. It allows the user to easily configure automations (which I guess is the sole point of such wireless switches). See

![device-trigger](https://github.com/user-attachments/assets/858d1bf8-b6c3-48d3-9f01-368ce0aad434)

Without this, users have to create automations by listening for `tuya_event` events, figure out the right values etc. which is very difficult for the average user.

Please let me know if I misunderstood your comment. I'm more than happy to iterate on this

frenck commented:
> The built-in Home Assistant Event component doesn't configure any device triggers itself as it doesn't understand the semantics of the events unfortunately in the current form.

That is not in scope for the integration, you are trying to work around a limitation in Home Assistant on the integration level.

Please remove them.

lanthaler commented:
OK, I removed the device triggers.

frenck commented on homeassistant/components/tuya/__init__.py:
```diff
@@ -134,6 +136,55 @@ async def async_remove_entry(hass: HomeAssistant, entry: TuyaConfigEntry) -> Non
 
 
+class TuyaManager(Manager):
+    """Patched tuya_sharing Manager.
+
+    This can be removed once change has been merged upstream in tuya-device-sharing-sdk. Please
+    check the status of https://github.com/tuya/tuya-device-sharing-sdk/pull/17
+    """
```
This is not code that should live in our codebase. We have requirements on things like these that they have to live in an separate package.

../Frenck

lanthaler replied:
Hmm... the last commit on https://github.com/tuya/tuya-device-sharing-sdk is from 9 months ago. None of the recent pull requests on that repo have received any response (https://github.com/tuya/tuya-device-sharing-sdk/pulls).

Which option do you suggest to move forward:

1. fork the entire code from https://github.com/tuya/tuya-device-sharing-sdk/tree/main/tuya_sharing into this component, i.e. move it under `homeassistant/components/tuya/...`
2. fork `tuya-device-sharing-sdk` and replace the dependency from this component with that new package
3. create a new library which just contains just this patch and add it as an additional dependency to this component (so it would have 2 dependencies going forward)

IMO 1) would be the best way forward as it allows the community here to maintain this code instead of continuing to rely on a unmaintained library.

If we go with 3), where should that code live and who should own that library?

frenck replied:
Did you contact Tuya?

frenck replied:
PS: The upstream PR you linked in the description is from yesterday... 🤷 

Like that is not much time you give to something open source.

lanthaler replied:
Yeah, I sent mine just yesterday but there are two others with one being a trivial 2 line change from November 2023 that hasn't been looked at. 

I haven't contacted Tuya as I don't have any contact there. Do you?

frenck commented:
Converted to draft yet again, as there are unresolved reviews.

lanthaler commented:
> Converted to draft yet again, as there are unresolved reviews.

I'm not familiar with the process here as this is my first HA PR. When should I flip it back to "ready for review"?

lanthaler commented:
@frenck I just rebased the PR onto the current dev. The patch of the Tuya library has also been removed as the changes landed upstream. Let me know if anything else needs to be changed.

Would be great if this could land in the next HA release :-)

danVnest commented:
Thanks for putting this together @lanthaler.
I have successfully been using this with two 'ja5osu5g' buttons (solving https://github.com/home-assistant/core/issues/120460) with no issues.

Let me know if I can help if any additional testing is required.

lanthaler commented:
@joostlek I see you approved #126180 which introduced a conflict with this PR.

I just rebased and resolved the merge conflict. Would you be able to review these changes? This PR has been sitting here for over a month now

frenck commented on homeassistant/components/tuya/manifest.json:
```diff
@@ -43,5 +43,5 @@
   "iot_class": "cloud_push",
   "loggers": ["tuya_iot"],
-  "requirements": ["tuya-device-sharing-sdk==0.1.9"]
+  "requirements": ["tuya-device-sharing-sdk==0.2.1"]
```
Dependency bumps need to go into their own separate PR. Please split this out from this pull request.

../Frenck

lanthaler replied:
OK. Split this off into #130333. I'll update this PR once that change has been merged

frenck reviewed (changes requested):
Besides the dependency bump comment below, the CI is currently not passing. Please take a look and that and make sure builds are all green.

Thanks! 👍 

../Frenck

lanthaler committed:
Add support for Tuya Wireless Switch entity

lanthaler commented:
I addressed all issues. All checks apart `codecov/patch` pass now. There seems to be more work required to properly set up tests for this integration. That should be done in a separate PR IMO.

frenck commented on homeassistant/components/tuya/strings.json:
```diff
@@ -101,6 +101,20 @@
+        "state_attributes": {
+          "event_type": {
+            "state": {
+              "click": "clicked",
+              "double_click": "double clicked",
+              "press": "long pressed"
```
```suggestion
              "click": "Clicked",
              "double_click": "Double-clicked",
              "press": "Long-pressed"
```

lanthaler replied:
Done

frenck commented on homeassistant/components/tuya/event.py:
```diff
@@ -0,0 +1,160 @@
+        self.entity_description = description
+        self._attr_unique_id = f"{super().unique_id}.{description.key}"
+
+        status_range = getattr(device, "status_range")
```
We have a `find_dpcode` helper method for that in the TuyaEntity class, please use that instead.

lanthaler replied:
Changed it

frenck commented on homeassistant/components/tuya/event.py:
```diff
@@ -0,0 +1,160 @@
+        value = self.device.status.get(self.entity_description.key)
+        self._trigger_event(value)
+        self.async_write_ha_state()
+        if TYPE_CHECKING:
+            assert self.device_entry
+            assert self.registry_entry
```
Not sure why this is added in the end?

```suggestion
```

lanthaler replied:
I also can't remember where this comes from. Removed it

frenck commented on homeassistant/components/tuya/event.py:
```diff
@@ -0,0 +1,160 @@
+        """Init Tuya event entity."""
+        super().__init__(device, device_manager)
+        self.entity_description = description
+        self._attr_unique_id = f"{super().unique_id}.{description.key}"
```
To match others:

```suggestion
        self._attr_unique_id = f"{super().unique_id}{description.key}"
```

lanthaler commented:
Thanks for the review @frenck. I addressed all feedback, squashed and rebased the branch onto dev.

lanthaler commented:
@frenck anything else that needs to change here or can this be merged now?

frenck reviewed (approved):
Nice! Thanks, @lanthaler 👍 

../Frenck

frenck reviewed (dismissed):
Hmm the CI is failing, could you take a look @lanthaler?

Thanks already! 👍 

../Frenck

lanthaler commented:
Adapted to upstream changes and rebased once more @frenck

lanthaler commented:
@frenck so for the ping but it has been almost 8 months by now since I opened this PR.

Any chance this could be merged now so that it makes it into the next release?

Thanks

emontnemery reviewed (approved):
LGTM, thanks @lanthaler 👍
